// waves-lfo/node/sink/DataToFile.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _from = require('babel-runtime/core-js/array/from');

var _from2 = _interopRequireDefault(_from);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _BaseLfo2 = require('../../core/BaseLfo');

var _BaseLfo3 = _interopRequireDefault(_BaseLfo2);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var definitions = {
  filename: {
    type: 'string',
    default: null,
    constant: true
  },
  format: {
    type: 'enum',
    default: 'txt',
    list: ['txt', 'json', 'csv']
  }
  // header: {
  //   type: 'boolean',
  //   default: false,
  // },
};

/**
 * Record input frames into a file.
 *
 * @param {Object} options - Override default parameters.
 * @param {String} [options.filename=null] - Path of the output file.
 * @param {String} [options.format='txt'] - Format in which the data should
 *  be stored. Available options: 'txt', 'json' or 'csv'.
 *
 * @memberof module:node.sink
 *
 * @todo - add an option to store the `streamParams`
 *
 * @example
 * import path from 'path';
 * import EventIn from '../src/common/source/EventIn';
 * import Logger from '../src/common/sink/Logger';
 * import DataToFile from '../src/node/sink/DataToFile';
 *
 * const eventIn = new EventIn({
 *   frameSize: 2,
 *   frameRate: 1,
 *   frameType: 'vector',
 * });
 *
 * const dataToFile = new DataToFile({
 *   filename: path.join(__dirname, './node_sink.DataToFile.test.json'),
 *   format: 'json',
 * });
 *
 * const logger = new Logger({
 *   data: true,
 * });
 *
 * eventIn.connect(logger);
 * eventIn.connect(dataToFile);
 * eventIn.start();
 *
 * let time = 0;
 * const period = 1;
 *
 * (function loop(){
 *   const data = [Math.random(), Math.random()];
 *   eventIn.process(time, data);
 *
 *   time += period;
 *
 *   if (time < 20)
 *     setTimeout(loop, 300);
 *   else
 *     eventIn.stop();
 * }());
 */

var DataToFile = function (_BaseLfo) {
  (0, _inherits3.default)(DataToFile, _BaseLfo);

  function DataToFile() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, DataToFile);

    var _this = (0, _possibleConstructorReturn3.default)(this, (DataToFile.__proto__ || (0, _getPrototypeOf2.default)(DataToFile)).call(this, definitions, options));

    _this.firstRow = true;
    return _this;
  }

  /** @private */


  (0, _createClass3.default)(DataToFile, [{
    key: 'processStreamParams',
    value: function processStreamParams(prevStreamParams) {
      this.prepareStreamParams(prevStreamParams);

      var filename = this.params.get('filename');
      this.file = _fs2.default.createWriteStream(filename);

      var format = this.params.get('format');

      switch (format) {
        case 'txt':
          break;
        case 'json':
          this.file.write('[\n');
          break;
        case 'csv':
          var _streamParams = this.streamParams,
              description = _streamParams.description,
              frameSize = _streamParams.frameSize;


          var header = 'time';

          if (description !== null && description.length) {
            header += ',' + description.join(',');
          } else {
            for (var i = 0; i < frameSize; i++) {
              header += ',row-' + i;
            }
          }

          header += '\n';

          this.file.write(header);
          break;
      }
    }

    /** @private */

  }, {
    key: 'finalizeStream',
    value: function finalizeStream(endTime) {
      var format = this.params.get('format');

      switch (format) {
        case 'txt':
          break;
        case 'json':
          this.file.write('\n]');
          break;
        case 'csv':
          // this.file.write(endTime.toString());
          break;
      }
    }

    // process any kind of stream
    /** @private */

  }, {
    key: 'processSignal',
    value: function processSignal() {}
    /** @private */

  }, {
    key: 'processVector',
    value: function processVector() {}
    /** @private */

  }, {
    key: 'processScalar',
    value: function processScalar() {}

    /** @private */

  }, {
    key: 'processFrame',
    value: function processFrame(frame) {
      var format = this.params.get('format');
      var str = void 0;

      switch (format) {
        case 'txt':
          str = _util2.default.format('%s\n', frame.data);
          break;
        case 'json':
          frame.data = (0, _from2.default)(frame.data);
          str = (0, _stringify2.default)(frame);

          if (!this.firstRow) str = ',\n' + str;

          this.firstRow = false;
          break;
        case 'csv':
          str = frame.time + ',';
          str += _util2.default.format('%s\n', frame.data);
          break;
      }

      this.file.write(str);
    }
  }]);
  return DataToFile;
}(_BaseLfo3.default);

exports.default = DataToFile;
